### PR TITLE
Clean up the OSD after the OSD is marked "out"

### LINKED ISSUE #2741: Clean up the OSD after the OSD is marked "out"
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
When an OSD is marked `out`, the operator currently only writes a message to the log that the OSD is down and takes no further action. When the OSD is out, we should consider removing the OSD:
- When the OSD is safe to remove (`ceph safe-to-remove`):
  - Delete the OSD deployment
  - Remove the OSD from the crush map
  - Turn on the blinking light (if available)

We should consider how aggressive we should be in this case. If a node was down temporarily and comes back online, the OSD should start again so that the PGs still assigned to the OSD won't have to be migrated. We would be covered fairly quickly if the "add node" event covered this scenario, but we wouldn't want to trigger an orchestration every time a node restarts either.

**What is use case behind this feature:**

**Environment**:

### LINKED ISSUE #3292: Log messages in Rook operator after node/OSD failure are unclear
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
When a Kubernetes node fails and Rook is waiting for the OSDs to come back online or to time out, the warning messages output to the operator log are unclear, and this has confused at least one user from Slack.

```
2019-06-12 20:53:09.332900 I | op-osd: osd.3 is marked 'DOWN'
2019-06-12 20:53:09.332914 W | op-osd: waiting for the osd.3 to exceed the grace period
```

**Expected behavior:**
The warning message is not very clear. It should state why Rook is waiting, what will happen after the grace period, what the grace period is, and how much time is left until the period is crossed. For example, ```OSD 3 is unavailable. Waiting for a grace period of GRACE_PERIOD seconds/minutes before assuming the OSD is gone forever, at which point the OSD will be removed from the Ceph cluster. TIME_REMAINING seconds/minutes remain in the grace period. It is natural to see the pod corresponding to this OSD waiting in "Pending" state at this time.```

When the OSD grace period is passed, I would expect the message to contain much of the same information. For example, ```OSD 3 has been unavailable for TIME_UNAVAILABLE seconds/minutes. This exceeds the allowed grace period of GRACE_PERIOD seconds/minutes. The OSD will now be removed from the Ceph cluster.```

It is also not clear whether Rook is observing its own grace period or whether it is allowing Ceph to mark the OSD `out` according to Ceph's own grace period.

**How to reproduce it (minimal and precise):**
Start a multi-node Rook cluster, and kill one of the nodes.

### PR DESCRIPTION
**Pull request by rohan47**
If for any reason an osd is marked out and ceph migrates its data and we determine it is safe-to-destroy, then rook would remove the K8s artifacts for those OSDs.

Signed-off-by: rohan47 <rohgupta@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #2741

[Addendum: Resolves #3292]

**Checklist:**
- [ ] Documentation has been updated, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](../CONTRIBUTING.md#comments)

// known CI issues
[skip ci]

### CODE REVIEW

**Comment by rohan47** [comment]
So far the code detects an OSD marked out, waits for the `osdGracePeriod` and then checks for `safe-to-destroy`
TODO: remove the K8s artifacts for those OSDs which are `safe-to-destroy`

**Comment by noahdesu** [comment]
meta question: since health monitoring (and I assume the eventual _actual_ removal of the OSD) runs asynchronously with orchestration, are there any races to worry about?

File: pkg/operator/ceph/cluster/osd/health.go [hunk 1]
```diff
@@ -73,14 +76,15 @@ func (m *Monitor) osdStatus() error {
...
+	evalDownStatus := func(id int, state string) {
 		if now := time.Now(); now.Sub(m.lastStatus[id]) > osdGracePeriod {
-			logger.Warningf("osd.%d has been down for longer than the grace period (down since %+v)", id, m.lastStatus[id])
+			logger.Warningf("osd.%d has been %v for longer than the grace period (%v since %+v)", id, state, state, m.lastStatus[id])
 			m.lastStatus[id] = time.Now()
 		} else {
 			logger.Warningf("waiting for the osd.%d to exceed the grace period", id)
 		}
 	}
+	evalOutStatus := evalDownStatus
```
[the remark this answers is not in the document]
Comment (rohan47) [reply pkg/operator/ceph/cluster/osd/health.go]: I made `evalOutStatus := evalDownStatus` just for readablity. `evalOutStatus` evaluates for osd in `out` state and  `evalDownStatus` evaluates for osd in `down` state. Was not able to figure out a common name need suggestions for that.

File: pkg/operator/ceph/cluster/osd/health.go [hunk 2]
```diff
@@ -23,7 +23,10 @@ import (
 	"github.com/rook/rook/pkg/daemon/ceph/client"
 )
 
-const upStatus = 1
+const (
+	upStatus = 1
+	inStatus = 1
+)
```
[the remark this answers is not in the document]
Comment (rohan47) [reply pkg/operator/ceph/cluster/osd/health.go]: Here `upStatus` and `inStatus` is used during comparision of output of `StatusByID` 
https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/osd/health.go#L96
https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/osd/health.go#L101

File: pkg/operator/ceph/cluster/osd/health.go [hunk 3]
```diff
@@ -112,7 +116,35 @@ func (m *Monitor) osdStatus() error {
 				delete(m.lastStatus, id)
 			}
 		}
+
+		if in != inStatus {
+			logger.Infof("osd.%d is marked 'OUT'", id)
+			if tracked {
+				evalOutStatus(id, "out")
+				if err := m.handleOSDMarkedOut(id); err != nil {
```
[the remark this answers is not in the document]
Comment (rohan47) [reply pkg/operator/ceph/cluster/osd/health.go]: There should be two different variables to track the timestamp for `down` and `out` as for the current scenario unit tests fails.

File: pkg/operator/ceph/cluster/osd/health.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (noahdesu) [reply pkg/operator/ceph/cluster/osd/health.go]: Yeh, and in `StatusByID` should `upStatus` and `inStatus` be returned rather than returning `1`?

File: pkg/operator/ceph/cluster/osd/health.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (noahdesu) [reply pkg/operator/ceph/cluster/osd/health.go]: You could probably do something like `evalOutStatus := func(id) { evalStatus(id, "out") }`

[the remark this answers is not in the document]
Comment (rohan47) [reply pkg/operator/ceph/cluster/osd/health.go]: updated the code @noahdesu

File: pkg/operator/ceph/cluster/osd/health.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (rohan47) [reply pkg/operator/ceph/cluster/osd/health.go]: I didn't get your point. Are you trying to say that `StatusByID` should return `upStatus` and `inStatus` when the return value is `1`?

File: pkg/operator/ceph/cluster/osd/health.go [hunk 4]
```diff
@@ -66,20 +70,24 @@ func (m *Monitor) Start(stopCh chan struct{}) {
...
 		return err
 	}
 	logger.Debugf("osd dump %v", osdDump)
 
-	evalDownStatus := func(id int) {
-		if now := time.Now(); now.Sub(m.lastStatus[id]) > osdGracePeriod {
-			logger.Warningf("osd.%d has been down for longer than the grace period (down since %+v)", id, m.lastStatus[id])
-			m.lastStatus[id] = time.Now()
+	//returns true if the laststatus is more than osd grace period else returns false
+	evalOutOrDownStatus := func(id int, state string, lastStatus map[int]time.Time) bool {
```
Comment (rohan47) [inline pkg/operator/ceph/cluster/osd/health.go]: returning a `bool` value in order to take action if the osdGracePeriod is exceeded for `down` or out` osds

File: pkg/operator/ceph/cluster/osd/health.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (noahdesu) [reply pkg/operator/ceph/cluster/osd/health.go]: `return upStatus` vs `return 1`

File: pkg/operator/ceph/cluster/osd/health.go [hunk 5]
```diff
@@ -37,15 +42,16 @@ type Monitor struct {
 
 	// lastStatus keeps track of OSDs status
 	// key - OSD id; value: time of the status change.
-	lastStatus map[int]time.Time
+	lastDownStatus map[int]time.Time
+	lastOutStatus  map[int]time.Time
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/health.go]: Do we need to track the out status? If an OSD is marked as `out`, we don't take any action until it is also `safe-to-destroy`, right? If it's out and safe to destroy, we can just remove it at that point rather than tracking in this map.

File: pkg/operator/ceph/cluster/osd/health.go [hunk 6]
```diff
@@ -66,20 +72,24 @@ func (m *Monitor) Start(stopCh chan struct{}) {
...
 	}
 	logger.Debugf("osd dump %v", osdDump)
 
-	evalDownStatus := func(id int) {
-		if now := time.Now(); now.Sub(m.lastStatus[id]) > osdGracePeriod {
-			logger.Warningf("osd.%d has been down for longer than the grace period (down since %+v)", id, m.lastStatus[id])
-			m.lastStatus[id] = time.Now()
+	//returns true if the laststatus is more than osd grace period else returns false
+	evalOutOrDownStatus := func(id int, state string, lastStatus map[int]time.Time) bool {
+		if now := time.Now(); now.Sub(lastStatus[id]) > osdGracePeriod {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/health.go]: Is the grace period needed anymore? I think we just need the following to keep it simple and there is no need to track grace periods in rook:
- Periodically query for down/out OSDs
- Log which OSDs are down/out
- For each OSD that is safe-to-destroy, clean it up

Comment (travisn) [reply pkg/operator/ceph/cluster/osd/health.go]: A long time in the past there was legacy code that did more with the `out` OSDs, but it really does seem like we can simplify this now.

**Comment by BlaineEXE** [comment]
Based on other discussions we have had around setting OSDs `noout` and when it is safe to remove OSDs, I am not sure if we *should* be working on this feature. An OSD being `out`, to my knowledge, doesn't mean that it is gone forever, and Rook should not remove the OSD prematurely.

Again I will defer to @liewegas and @leseb for their thoughts about whether this could be disrupting to admins.

**Comment by travisn** [comment]
The discussion was captured [here](https://github.com/rook/rook/issues/2741#issuecomment-479998478). This PR is to handle the second bullet which is to remove the k8s artifacts only when the osd is safe-to-remove. @BlaineEXE Is there some other follow up discussion I'm forgetting?

**Comment by BlaineEXE** [comment]
> If for any reason an osd is marked out and ceph migrates its data and we determine it is safe-to-remove, then rook would remove the K8s artifacts for those OSDs. (This case is the original intention of this issue)

It's not clear to me that this is a good procedure from a cluster admin perspective. As an example, let's assume that a node or disk controller fails, and it takes some time for the admin to get a replacement. In the meantime, Ceph will rebalance the data and then Ceph may mark the osds as safe to delete (which versions of Ceph do this?). But what happens when the node is later repaired? The old disks are fine and still have some data on them. If we don't remove the OSDs, can Ceph pick them back up when they're online and get back to ideal state without having to do too much data rebalancing? 

If we implement this feature in Rook, we would be forced to create brand new OSDs on the repaired node after wiping all the disks, so a full 1-node rebalance would be necessary in this case. It's my understanding these rebalances can take a significant amount of time. There may be a case to be made for Rook being less aggressive about removing disks to speed up recovery times (assuming this is possible) in cases like this. I really don't know, but I would want a better understanding of ideal administrator workflows before proceeding.

**Comment by noahdesu** [comment]
I'm still on the fence about this. Primarily because I can imagine why it would be convenient to automatically remove OSDs (after all rook is supposed to fill at least part of the role of a cluster administrator), but I haven't seen a compelling reason to automatically remove OSDs when weighed against the open questions concerning the negatives of automatic removal.

When driven by administrative tasks such as explicitly removing a k8s node (or any related implicit changes that k8s may make to the source of truth cluster information that affects ceph), automatically passing along those administrative tasks to remove OSDs, etc... makes sense. So what follows is only related to automatic OSD removal assuming an otherwise static k8s cluster configuration.

I believe the current proposal is to automatically remove OSDs that enter the safe-to-destroy state. So this would imply they are out due to a number of reasons such as (1) disk failure (2) node failure (3) power failure (4) network failure/partition (5) other transient (6) other perm failure, etc...

Of these causes, if perm failures (like disk failures) could be reliably detected that might be a reason for automatic OSD removal, but would administrators be keen to replace the disk to restore balance between failure domains (leaving the OSD in crush map), and what are the benefits to replacing disks by removing the OSD ID?

Other causes might be transient and correlated (e.g. network partition) so answering this question for multiple OSDs compared to a single OSD seems helpful. Removing a whole rack of nodes for a temporary failure doesn't seem terribly efficient.

In summary, from my perspective it boils down to: does automatic removal of `out` osds provide some tangible non-administrative benefit (e.g. performance), and what are the impacts / likelihood of all the other possibilities? Would using a very large timeout (e.g. 1-2 days) make sense?

**Comment by travisn** [comment]
If the OSDs are both `out` and `safe-to-destroy`, that means whatever data was held by the OSDs has been backfilled elsewhere. If the OSD comes back online, it won't be considered to be holding any data. If this is the case:
- Removing the OSD does not cause any harm. It's just cleanup of K8s resources. 
- If the disk comes back online, the operator should see that event and start the osd pod again. The osd just wouldn't be holding any data anymore. This event could cause rebalancing, the same as adding any new OSD to the cluster.
Again, the critical decision point here is that it's safe-to-destroy. Otherwise, Rook shouldn't take action to remove an osd.

**Comment by liewegas** [comment]
If an OSD is marked out and is then removed from the CRUSH map, it triggers data reshuffling.  So all things being equal I'd lean toward *not* taking any action.  That leaves the door open for the (presumably failed?) device to be replaced in place and for the resulting recovery to move data only once (not twice).

**Comment by travisn** [comment]
To be clear, the OSD is not automatically removed from the CRUSH map by Ceph, right? I don't see that Rook should remove it from the CRUSH map automatically either. This change should only be about cleaning up the K8s pod/resources when an osd is safe-to-destroy. The follow-on scenarios would be:
- The device comes back online. Rook will see it's already configured and create its pod again.
- The device is replaced and looks like a new device to Rook. Rook will call `c-v` to provision an OSD and then start the daemon. 
Either way, the old OSD ID would remain in the CRUSH map, correct? It would just be marked as `out` indefinitely?

**Comment by liewegas** [comment]
> To be clear, the OSD is not automatically removed from the CRUSH map by Ceph, right? I don't see that Rook should remove it from the CRUSH map automatically either. This change should only be about cleaning up the K8s pod/resources when an osd is safe-to-destroy.

Ah, right.  It's the ``ceph osd purge`` command that cleans the OSD out of the crush map (and removes the auth keys).

> The follow-on scenarios would be:
> 
> * The device comes back online. Rook will see it's already configured and create its pod again.
> * The device is replaced and looks like a new device to Rook. Rook will call `c-v` to provision an OSD and then start the daemon.
>   Either way, the old OSD ID would remain in the CRUSH map, correct? It would just be marked as `out` indefinitely?

Right.  And in the latter case, if rook can infer that this device is replacing the old osd's device (this may be a bit tricky, but doable), then c-v can be passed the old osd-id so that it is reused.  This is the "correct"/best replacement scenario.

**Comment by travisn** [comment]
Ah that's the piece I was missing to replace the OSD, it's as simple as passing the old osd id to c-v.

**Comment by noahdesu** [comment]
> This change should only be about cleaning up the K8s pod/resources when an osd is safe-to-destroy. 

Thanks for clarification

File: pkg/operator/ceph/cluster/osd/health.go [hunk 7]
```diff
@@ -90,29 +82,60 @@ func (m *Monitor) osdStatus() error {
...
-			if tracked {
-				logger.Infof("osd.%d recovered, stopping tracking.", id)
-				delete(m.lastStatus, id)
-			}
+
+		}
+
+		if in != inStatus {
+			logger.Infof("osd.%d is marked 'OUT'", id)
+			m.outOSDs[strconv.Itoa(id)] = true
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/health.go]: Do we need to save the `out` OSDs into the map? We can just immediately call `handleOSDMarkedOut()` and remove the K8s resources, right? I don't think we need to batch these at the same time for all `out` OSDs.

File: pkg/operator/ceph/cluster/osd/health.go [hunk 8]
```diff
@@ -90,29 +82,60 @@ func (m *Monitor) osdStatus() error {
...
+	safeToDestroyOSDs, err := client.OsdSafeToDestroy(m.context, m.clusterName, outOSDids)
+	if err != nil {
+		return err
+	}
+
+	for _, id := range safeToDestroyOSDs.SafeToDestroy {
+		logger.Infof("osd.%d is 'safe-to-destroy'", id)
+		label := fmt.Sprintf("ceph-osd-id=%d", id)
+		dp, err := k8sutil.GetDeployments(m.context.Clientset, m.clusterName, label)
+		if err != nil {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/health.go]: A valid error could be that the deployment is not found. In fact, we would expect this frequently. After an OSD has been removed by this code path previously, the OSD will look like it needs to be removed every time this check is done. So we will need to ignore the error in that case.

File: pkg/operator/ceph/cluster/osd/health.go (hunk 7 above)
Comment (rohan47) [reply pkg/operator/ceph/cluster/osd/health.go]: `ceph osd safe-to-destroy` command can take multiple osd ids and return the osd ids that are `safe-to-destroy`. So what I was thinking was to utilize this. After every 5 min make a map of all the out osds and pass them to `safe-to-destroy` command (i.e., to `OsdSafeToDestroy()` function). Also when the output of `OsdSafeToDestroy()` is deserialized into json it gives a struct having a list of int(list of osd ids that are safe to destroy).

File: pkg/operator/ceph/cluster/osd/health.go (hunk 8 above)
Comment (rohan47) [reply pkg/operator/ceph/cluster/osd/health.go]: Got it.

File: pkg/daemon/ceph/client/osd.go [hunk 9]
```diff
@@ -150,6 +154,22 @@ func OSDRemove(context *clusterd.Context, clusterName string, osdID int) (string
 	return string(buf), err
 }
 
+func OsdSafeToDestroy(context *clusterd.Context, clusterName string, osdID []string) (*SafeToDestroyStatus, error) {
```
Comment (rohan47) [inline pkg/daemon/ceph/client/osd.go]: `ceph osd safe-to-destroy` can take multiple osd ids and return all the osds that are safe to destroy. So i took `osdID` as a `[]string`. So will it be fine if I take just the `osdID` as `int` or `string`, even though the output of the command after deserialization is a struct containing `[]int` ?

File: pkg/operator/ceph/cluster/osd/health.go (hunk 7 above)
Comment (rohan47) [reply pkg/operator/ceph/cluster/osd/health.go]: I saved the `out` OSDs into map based on https://github.com/rook/rook/pull/3117#discussion_r284426435

Comment (travisn) [reply pkg/operator/ceph/cluster/osd/health.go]: Yeah, now that I see how simple we can make the code path, I'm ready to change the recommendation. It should be a rare operation to remove an OSD after it is safe to remove, so I don't expect we will gain anything by batching the operation. Simplifying the code is always my preference if we can. :)

Comment (rohan47) [reply pkg/operator/ceph/cluster/osd/health.go]: cool :+1:

File: pkg/daemon/ceph/client/osd.go (hunk 9 above)
Comment (travisn) [reply pkg/daemon/ceph/client/osd.go]: Yes, sounds good to just add a single osd as input, but then you would need to handle the output as an array.

**Comment by BlaineEXE** [comment]
A lot of users want to have a hands-off experience with their cluster and don't want to have to manually remove OSDs. If we remove pods but don't remove OSDs from the CRUSH map for these users, are we doing them a disservice? They will be very confused by their Ceph cluster reporting that there are OSDs missing.

I think we should always keep pods around if Ceph still has them in its map. The pods will be stuck in "Pending" state if they are failed, but this could be a good indicator to the user that things are not 100% well.

What if we had a configuration parameter that allowed Rook to remove OSDs automatically? Something like `storage:automaticOSDRemoval: true/false` would make sense I think. If this parameter is `true`, then Rook purges the OSD from the CRUSH map and deletes the deployments. If it is false, Rook should keep the pods.

We could consider keeping the deployments but scale the pods to 0 (zero) replicas if automatic OSD removal is `false`. Rook would still have a record of the deployment (because it still exists), but users wouldn't be distressed by "Pending" pods as one of the slack users has been struggling with the past couple days.

**Comment by BlaineEXE** [comment]
@travisn I don't see anything wrong with that workflow. That seems pretty accurate to me. I'm okay with this going in.

File: pkg/operator/ceph/cluster/osd/health.go [hunk 10]
```diff
@@ -90,29 +80,46 @@ func (m *Monitor) osdStatus() error {
...
-			if tracked {
-				logger.Infof("osd.%d recovered, stopping tracking.", id)
-				delete(m.lastStatus, id)
+
+		}
+
+		if in != inStatus {
+			logger.Debugf("osd.%d is marked 'OUT'", id)
+			if err := m.handleOSDMarkedOut(id); err != nil {
+				return err
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/health.go]: How about if we just log this error instead of return it? This is best effort to clean up the OSDs. A single failure to cleanup shouldn't block other OSDs from cleaning up.

Comment (rohan47) [reply pkg/operator/ceph/cluster/osd/health.go]: Updated

File: pkg/daemon/ceph/client/osd.go [hunk 11]
```diff
@@ -150,6 +154,24 @@ func OSDRemove(context *clusterd.Context, clusterName string, osdID int) (string
 	return string(buf), err
 }
 
+func OsdSafeToDestroy(context *clusterd.Context, clusterName string, osdID int) (bool, error) {
+	args := []string{"osd", "safe-to-destroy", strconv.Itoa(osdID)}
+	cmd := NewCephCommand(context, clusterName, args)
```
Comment (leseb) [inline pkg/daemon/ceph/client/osd.go]: It's not 100% safe since before Nautilus this command didn't have any json output... A version check would have been nice...

Comment (rohan47) [reply pkg/daemon/ceph/client/osd.go]: Okay, Thanks, Will update

Comment (travisn) [reply pkg/daemon/ceph/client/osd.go]: @rohan47 Is there an issue opened to track this? thanks

Comment (rohan47) [reply pkg/daemon/ceph/client/osd.go]: No, but I will fix it with https://github.com/rook/rook/issues/3359

Comment (rohan47) [reply pkg/daemon/ceph/client/osd.go]: @travisn @leseb I have opened a separate issue for this https://github.com/rook/rook/issues/3421
